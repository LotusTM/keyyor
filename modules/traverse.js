/**
 * Take any input, traverse it and apply specified function recursively
 * @param {*}        input Input which should be traversed
 * @param {function} fn    Function to be applied for each matching leaf
 *                         Receives matching predicate input as first argument
 * @param {function} [predicate = () => true]
 *                         Function to select matching inputs, receives input as first arg
 *                         Should return boolean
 * @return {*} Cloned and transformed input
 * @example
 *  traverse(
 *    { apply: 'APPLY', inner: { apply: 'APPLY' } },
 *    (tmpl) => tmpl + ' fn APPLIED',
 *    (string) => string.includes('APPLY')
 *  )
 *  // -> { apply: 'APPLY fn APPLIED', inner: { apply: 'APPLY fn APPLIED' } }
*/
const traverse = (input, fn, predicate = () => true) => {
  const tryFn = (input) => {
    try {
      return fn(input)
    } catch (e) {
      e.message = `\n\n\`\`\`\n${input}\n\n\`\`\`\n\n${e.message}`
      throw e
    }
  }

  if (input === undefined || input === null) return input

  // For cases when recevied String Object instead of primitive from Nunjucks macro
  input = input.valueOf()

  if (typeof input === 'string') return predicate(input) ? tryFn(input) : input

  if (typeof input === 'object') {
    const clone = Array.isArray(input) ? input.slice(0) : Object.assign({}, input)

    for (let key in input) {
      clone[key] = traverse(input[key], fn, predicate)
    }

    return clone
  }

  return input
}

module.exports = traverse
