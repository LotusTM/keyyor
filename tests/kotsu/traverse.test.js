/* eslint-env jest */

import traverse from '../../modules/traverse'

describe('Traverse function', () => {
  const fn = (tmpl) => tmpl + ' fn APPLIED'
  const predicate = (string) => string.includes('APPLY')

  it('should traverse string', () => {
    expect(traverse('APPLY', fn, predicate)).toMatchSnapshot()
  })

  it('should traverse object', () => {
    expect(traverse({
      apply: 'APPLY apply',
      notApply: 'nope',
      und: undefined,
      nll: null,
      true: true,
      false: false,
      num: 123,
      inner: {
        apply: 'APPLY apply',
        notApply: 'nope',
        und: undefined,
        nll: null,
        true: true,
        false: false,
        num: 123
      }
    }, fn, predicate)).toMatchSnapshot()
  })

  it('should traverse array', () => {
    expect(traverse([
      'APPLY apply', 'nope', undefined, null, false, true, 123,
      ['APPLY apply', 'nope', undefined, null, false, true, 123]
    ], fn, predicate)).toMatchSnapshot()
  })

  it('should print `input` in error message', () => {
    expect(() => traverse(
      ['not triggers', 'triggers error'],
      (input) => {
        if (input.includes('error')) throw new SyntaxError('String should not contain `error`')
      }
    )).toThrowErrorMatchingSnapshot()
  })
})
