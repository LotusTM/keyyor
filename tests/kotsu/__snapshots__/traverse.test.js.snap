// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traverse function should print \`input\` in error message 1`] = `
"

\`\`\`
triggers error

\`\`\`

String should not contain \`error\`"
`;

exports[`Traverse function should traverse array 1`] = `
Array [
  "APPLY apply fn APPLIED",
  "nope",
  undefined,
  null,
  false,
  true,
  123,
  Array [
    "APPLY apply fn APPLIED",
    "nope",
    undefined,
    null,
    false,
    true,
    123,
  ],
]
`;

exports[`Traverse function should traverse object 1`] = `
Object {
  "apply": "APPLY apply fn APPLIED",
  "false": false,
  "inner": Object {
    "apply": "APPLY apply fn APPLIED",
    "false": false,
    "nll": null,
    "notApply": "nope",
    "num": 123,
    "true": true,
    "und": undefined,
  },
  "nll": null,
  "notApply": "nope",
  "num": 123,
  "true": true,
  "und": undefined,
}
`;

exports[`Traverse function should traverse string 1`] = `"APPLY fn APPLIED"`;
